File: Data Structures & Algorithms/sort-colors/submission-2.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def swap(nums, l, r):
            nums[l], nums[r] = nums[r], nums[l]

        low, mid, high = 0, 0, len(nums) - 1

        while mid <= high:
            val = nums[mid]
            if val == 0:
                swap(nums, low, mid)
                low += 1
                mid += 1
            elif val == 1:
                mid += 1
            else:
                swap(nums, mid, high)
                high -= 1

File: Data Structures & Algorithms/binary-search/submission-2.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target, int st, int end) {
        if (st > end) return -1;

        int mid = st + (end - st)/2;
        if (nums[mid] < target) {
            return binarySearch(nums, target, mid + 1, end);
        } else if (nums[mid] > target) {
            return binarySearch(nums, target, st, end - 1);
        } else {
            return mid;
        }
    }

    int search(vector<int>& nums, int target) {
        return binarySearch(nums, target, 0, nums.size() - 1);
    }
};
